File: articles/lifespan-developmental-psychology-and-the-baltes-tradition/go/lifespan_baltes_summary.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"
)

func col(header []string, name string) int {
	for i, h := range header {
		if h == name {
			return i
		}
	}
	return -1
}

func main() {
	if len(os.Args) < 2 {
		log.Fatal("Usage: go run lifespan_baltes_summary.go data/lifespan_baltes_panel.csv")
	}
	f, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	rows, err := csv.NewReader(f).ReadAll()
	if err != nil {
		log.Fatal(err)
	}
	if len(rows) < 2 {
		log.Fatal("CSV has no data rows")
	}

	header := rows[0]
	developmentIdx := col(header, "development_score")
	gainsIdx := col(header, "gains")
	lossesIdx := col(header, "losses")
	socIdx := col(header, "soc_index")
	supportIdx := col(header, "current_support")
	if developmentIdx < 0 || gainsIdx < 0 || lossesIdx < 0 || socIdx < 0 || supportIdx < 0 {
		log.Fatal("Required columns missing")
	}

	count := 0
	developmentSum, gainsSum, lossesSum, socSum, supportSum := 0.0, 0.0, 0.0, 0.0, 0.0

	for _, row := range rows[1:] {
		development, e1 := strconv.ParseFloat(row[developmentIdx], 64)
		gains, e2 := strconv.ParseFloat(row[gainsIdx], 64)
		losses, e3 := strconv.ParseFloat(row[lossesIdx], 64)
		soc, e4 := strconv.ParseFloat(row[socIdx], 64)
		support, e5 := strconv.ParseFloat(row[supportIdx], 64)
		if e1 != nil || e2 != nil || e3 != nil || e4 != nil || e5 != nil {
			continue
		}
		count++
		developmentSum += development
		gainsSum += gains
		lossesSum += losses
		socSum += soc
		supportSum += support
	}

	fmt.Printf("Rows analyzed: %d\n", count)
	fmt.Printf("Mean development_score: %.4f\n", developmentSum/float64(count))
	fmt.Printf("Mean gains: %.4f\n", gainsSum/float64(count))
	fmt.Printf("Mean losses: %.4f\n", lossesSum/float64(count))
	fmt.Printf("Mean soc_index: %.4f\n", socSum/float64(count))
	fmt.Printf("Mean current_support: %.4f\n", supportSum/float64(count))
}

File: articles/lifespan-developmental-psychology-and-the-baltes-tradition/c/lifespan_baltes_summary.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINE 16384

int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s data/lifespan_baltes_panel.csv\n", argv[0]);
        return 1;
    }

    FILE *fp = fopen(argv[1], "r");
    if (!fp) {
        perror("Unable to open CSV file");
        return 1;
    }

    char line[MAX_LINE];
    if (!fgets(line, sizeof(line), fp)) {
        fprintf(stderr, "CSV is empty\n");
        fclose(fp);
        return 1;
    }

    int development_col = -1, gains_col = -1, losses_col = -1, soc_col = -1, support_col = -1, col = 0;
    char *token = strtok(line, ",");

    while (token != NULL) {
        token[strcspn(token, "\r\n")] = 0;
        if (strcmp(token, "development_score") == 0) development_col = col;
        if (strcmp(token, "gains") == 0) gains_col = col;
        if (strcmp(token, "losses") == 0) losses_col = col;
        if (strcmp(token, "soc_index") == 0) soc_col = col;
        if (strcmp(token, "current_support") == 0) support_col = col;
        token = strtok(NULL, ",");
        col++;
    }

    if (development_col < 0 || gains_col < 0 || losses_col < 0 || soc_col < 0 || support_col < 0) {
        fprintf(stderr, "Required columns not found\n");
        fclose(fp);
        return 1;
    }

    long count = 0;
    double development_sum = 0.0, gains_sum = 0.0, losses_sum = 0.0, soc_sum = 0.0, support_sum = 0.0;

    while (fgets(line, sizeof(line), fp)) {
        int current_col = 0;
        double development = 0.0, gains = 0.0, losses = 0.0, soc = 0.0, support = 0.0;
        char *field = strtok(line, ",");

        while (field != NULL) {
            if (current_col == development_col) development = atof(field);
            if (current_col == gains_col) gains = atof(field);
            if (current_col == losses_col) losses = atof(field);
            if (current_col == soc_col) soc = atof(field);
            if (current_col == support_col) support = atof(field);
            field = strtok(NULL, ",");
            current_col++;
        }

        count++;
        development_sum += development;
        gains_sum += gains;
        losses_sum += losses;
        soc_sum += soc;
        support_sum += support;
    }

    fclose(fp);

    printf("Rows analyzed: %ld\n", count);
    printf("Mean development_score: %.4f\n", development_sum / count);
    printf("Mean gains: %.4f\n", gains_sum / count);
    printf("Mean losses: %.4f\n", losses_sum / count);
    printf("Mean soc_index: %.4f\n", soc_sum / count);
    printf("Mean current_support: %.4f\n", support_sum / count);
    return 0;
}
